guava/src/com/google/common/escape/UnicodeEscaper.java

```diff
@@ -242,16 +259,17 @@ protected final String escapeSlow(String s, int index) {
    * @return the Unicode code point for the given index or the negated value of the trailing high
    *     surrogate character at the end of the sequence
    */
-  protected static int codePointAt(CharSequence seq, int index, int end) {
+  protected static int codePointAt(CharSequence seq, @IndexFor("#1") int index, @IndexOrHigh("#1") int end) {
     checkNotNull(seq);
-    if (index < end) {
-      char c1 = seq.charAt(index++);
+    @IndexOrHigh("seq") int indexInternal = index;
+    if (indexInternal < end) {
+      char c1 = seq.charAt(indexInternal++);
       if (c1 < Character.MIN_HIGH_SURROGATE || c1 > Character.MAX_LOW_SURROGATE) {
         // Fast path (first test is probably all we need to do)
         return c1;
       } else if (c1 <= Character.MAX_HIGH_SURROGATE) {
         // If the high surrogate was the last character, return its inverse
-        if (index == end) {
+        if (indexInternal == end) {
           return -c1;
         }
         // Otherwise look for the low surrogate following it
```

Comment:
the exceptions below should use the value of `indexInternal` in their messages, not the value of `index`

Reply:
I made changes for this comment already, but since you commented in the `before changes` version, this comment just won't close.